File: sdks/python/src/openspawn/resources/credits.py
"""Credits resource module."""

from typing import Any, Optional

from openspawn.http_client import AsyncHTTPClient, SyncHTTPClient
from openspawn.models import CreditTransaction, TransferCreditsRequest


class CreditsResource:
    """Synchronous credits resource."""

    def __init__(self, client: SyncHTTPClient) -> None:
        self._client = client

    def balance(self) -> int:
        """
        Get current agent's credit balance.

        Returns:
            Current balance
        """
        response = self._client.request("GET", "/credits/balance")
        return response["data"]

    def transfer(self, request: TransferCreditsRequest) -> CreditTransaction:
        """
        Transfer credits to another agent.

        Args:
            request: Transfer request

        Returns:
            Credit transaction
        """
        response = self._client.request(
            "POST",
            "/agents/credits/transfer",
            json_data=request.model_dump(by_alias=True),
            idempotent=True,
        )
        return CreditTransaction(**response["data"])

    def history(
        self, *, limit: int = 50, offset: int = 0
    ) -> tuple[list[CreditTransaction], dict[str, Any]]:
        """
        Get credit transaction history for current agent.

        Args:
            limit: Maximum number of transactions to return
            offset: Number of transactions to skip

        Returns:
            Tuple of (transactions list, metadata dict)
        """
        response = self._client.request(
            "GET",
            "/credits/history",
            params={"limit": str(limit), "offset": str(offset)},
        )
        transactions = [CreditTransaction(**tx) for tx in response["data"]]
        return transactions, response.get("meta", {})

    def spend(
        self,
        amount: int,
        reason: str,
        *,
        trigger_type: Optional[str] = None,
        source_task_id: Optional[str] = None,
        source_agent_id: Optional[str] = None,
    ) -> CreditTransaction:
        """
        Spend credits (record a debit transaction).

        Args:
            amount: Amount to spend
            reason: Reason for spending
            trigger_type: Type of trigger
            source_task_id: Related task ID
            source_agent_id: Related agent ID

        Returns:
            Credit transaction
        """
        response = self._client.request(
            "POST",
            "/credits/spend",
            json_data={
                "amount": amount,
                "reason": reason,
                "triggerType": trigger_type,
                "sourceTaskId": source_task_id,
                "sourceAgentId": source_agent_id,
            },
            idempotent=True,
        )
        return CreditTransaction(**response["data"])


class AsyncCreditsResource:
    """Asynchronous credits resource."""

    def __init__(self, client: AsyncHTTPClient) -> None:
        self._client = client

    async def balance(self) -> int:
        """
        Get current agent's credit balance.

        Returns:
            Current balance
        """
        response = await self._client.request("GET", "/credits/balance")
        return response["data"]

    async def transfer(self, request: TransferCreditsRequest) -> CreditTransaction:
        """
        Transfer credits to another agent.

        Args:
            request: Transfer request

        Returns:
            Credit transaction
        """
        response = await self._client.request(
            "POST",
            "/agents/credits/transfer",
            json_data=request.model_dump(by_alias=True),
            idempotent=True,
        )
        return CreditTransaction(**response["data"])

    async def history(
        self, *, limit: int = 50, offset: int = 0
    ) -> tuple[list[CreditTransaction], dict[str, Any]]:
        """
        Get credit transaction history for current agent.

        Args:
            limit: Maximum number of transactions to return
            offset: Number of transactions to skip

        Returns:
            Tuple of (transactions list, metadata dict)
        """
        response = await self._client.request(
            "GET",
            "/credits/history",
            params={"limit": str(limit), "offset": str(offset)},
        )
        transactions = [CreditTransaction(**tx) for tx in response["data"]]
        return transactions, response.get("meta", {})

    async def spend(
        self,
        amount: int,
        reason: str,
        *,
        trigger_type: Optional[str] = None,
        source_task_id: Optional[str] = None,
        source_agent_id: Optional[str] = None,
    ) -> CreditTransaction:
        """
        Spend credits (record a debit transaction).

        Args:
            amount: Amount to spend
            reason: Reason for spending
            trigger_type: Type of trigger
            source_task_id: Related task ID
            source_agent_id: Related agent ID

        Returns:
            Credit transaction
        """
        response = await self._client.request(
            "POST",
            "/credits/spend",
            json_data={
                "amount": amount,
                "reason": reason,
                "triggerType": trigger_type,
                "sourceTaskId": source_task_id,
                "sourceAgentId": source_agent_id,
            },
            idempotent=True,
        )
        return CreditTransaction(**response["data"])

File: libs/sdk/src/lib/resources/credits.ts
import type { HttpClient } from '../http-client';
import type { ApiResponse, RequestOptions } from '../types';
import type { CreditType } from '../shared-types';

/**
 * Credit-related types
 */
export interface CreditBalance {
  balance: number;
  agentId: string;
}

export interface CreditTransaction {
  id: string;
  orgId: string;
  agentId: string;
  amount: number;
  type: CreditType;
  reason: string;
  balanceBefore: number;
  balanceAfter: number;
  sourceTaskId?: string;
  sourceAgentId?: string;
  triggerType?: string;
  createdAt: string;
}

export interface SpendCreditsDto {
  amount: number;
  reason: string;
  triggerType?: string;
  sourceTaskId?: string;
  sourceAgentId?: string;
}

export interface TransferCreditsDto {
  toAgentId: string;
  amount: number;
  reason: string;
}

export interface CreditHistoryResponse {
  transactions: CreditTransaction[];
  total: number;
  limit: number;
  offset: number;
}

/**
 * Credits Resource
 */
export class CreditsResource {
  constructor(private http: HttpClient) {}

  /**
   * Get current credit balance for authenticated agent
   */
  async balance(options?: RequestOptions): Promise<number> {
    const response = await this.http.get<ApiResponse<CreditBalance>>(
      '/credits/balance',
      options
    );
    return response.data.balance;
  }

  /**
   * Spend credits
   */
  async spend(
    data: SpendCreditsDto,
    options?: RequestOptions
  ): Promise<CreditTransaction> {
    const response = await this.http.post<ApiResponse<CreditTransaction>>(
      '/credits/spend',
      data,
      options
    );
    return response.data;
  }

  /**
   * Get credit transaction history
   */
  async history(
    limit = 50,
    offset = 0,
    options?: RequestOptions
  ): Promise<CreditHistoryResponse> {
    const response = await this.http.get<
      ApiResponse<CreditTransaction[]> & {
        meta: { total: number; limit: number; offset: number };
      }
    >(`/credits/history?limit=${limit}&offset=${offset}`, options);

    return {
      transactions: response.data,
      total: response.meta.total,
      limit: response.meta.limit,
      offset: response.meta.offset,
    };
  }

  /**
   * Transfer credits to another agent
   */
  async transfer(
    data: TransferCreditsDto,
    options?: RequestOptions
  ): Promise<unknown> {
    const response = await this.http.post<ApiResponse<unknown>>(
      '/agents/credits/transfer',
      data,
      options
    );
    return response.data;
  }

  /**
   * Get organization-wide credit statistics
   */
  async getStats(options?: RequestOptions): Promise<unknown> {
    const response = await this.http.get<ApiResponse<unknown>>(
      '/credits/analytics/stats',
      options
    );
    return response.data;
  }

  /**
   * Get spending trends over time
   */
  async getTrends(
    days = 30,
    agentId?: string,
    options?: RequestOptions
  ): Promise<unknown> {
    const params = new URLSearchParams({ days: String(days) });
    if (agentId) params.set('agentId', agentId);

    const response = await this.http.get<ApiResponse<unknown>>(
      `/credits/analytics/trends?${params.toString()}`,
      options
    );
    return response.data;
  }

  /**
   * Get spending summary by agent
   */
  async getAgentSummary(options?: RequestOptions): Promise<unknown> {
    const response = await this.http.get<ApiResponse<unknown>>(
      '/credits/analytics/agents',
      options
    );
    return response.data;
  }

  /**
   * Get top spenders
   */
  async getTopSpenders(
    days = 7,
    limit = 10,
    options?: RequestOptions
  ): Promise<unknown> {
    const params = new URLSearchParams({
      days: String(days),
      limit: String(limit),
    });

    const response = await this.http.get<ApiResponse<unknown>>(
      `/credits/analytics/top-spenders?${params.toString()}`,
      options
    );
    return response.data;
  }
}
